README.md
# haveibeenpwned

[![GitHub release](http://img.shields.io/github/release/jakewarren/haveibeenpwned.svg?style=flat-square)](https://github.com/jakewarren/haveibeenpwned/releases])
[![CircleCI](https://circleci.com/gh/jakewarren/haveibeenpwned.svg?style=shield)](https://circleci.com/gh/jakewarren/haveibeenpwned)
[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](https://github.com/jakewarren/haveibeenpwned/blob/master/LICENSE)
[![Go Report Card](https://goreportcard.com/badge/github.com/jakewarren/haveibeenpwned)](https://goreportcard.com/report/github.com/jakewarren/haveibeenpwned)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=shields)](http://makeapullrequest.com)
[![GoDoc](https://img.shields.io/badge/godoc-reference-blue.svg)](https://godoc.org/github.com/jakewarren/haveibeenpwned/api)

> library and cmd-line client for the HIBP API

## Install
### Option 1: Binary

Download the latest release from [https://github.com/jakewarren/haveibeenpwned/releases/latest](https://github.com/jakewarren/haveibeenpwned/releases/latest)

### Option 2: From source

```
go get github.com/jakewarren/haveibeenpwned
```

## Usage

```
❯ haveibeenpwned -h
usage: haveibeenpwned [<flags>] <email>

Un-official API client for haveibeenpwned.com.

Optional flags:
  -h, --help                     Show context-sensitive help (also try --help-long and --help-man).
  -d, --debug                    print debug info
  -f, --filter-date=FILTER-DATE  only print breaches released after specified date
  -s, --silent                   suppress response message, only display results
  -V, --version                  Show application version.

Args:
  <email>  the email address to lookup.
```

## Changes

All notable changes to this project will be documented in the [changelog].

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## Example of using the library

```
package main

import (

	"fmt"
	"log"
    "github.com/jakewarren/haveibeenpwned/api"

)

func main() {

    // initialize the client with your API token
    client = api.NewClient("abc123")

	breachData,err := client.LookupEmailBreaches("test@example.com")
	
	if err != nil {
		log.Fatal(err)
	}


	for _,breach := range breachData {
		fmt.Printf("Breach: %s\n",breach.Title)
	}
}
```

## Thanks to :heart:

* [@MasenkoHa](https://github.com/MasenkoHa)

## License

MIT © 2018 Jake Warren

[changelog]: https://github.com/jakewarren/haveibeenpwned/blob/master/CHANGELOG.md


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.1] - 2019-11-18
### Changed
- Migrated to go modules
- Fixed up misc linting issues

## [2.0.0] - 2019-09-04
- Updated to support v3 of the HIBP API

## [1.0.0] - 2018-03-28
- Initial Release


[unreleased]: https://github.com/jakewarren/haveibeenpwned/compare/v2.0.1...HEAD
[2.0.1]: https://github.com/jakewarren/haveibeenpwned/compare/v0.1.2...v2.0.1
[2.0.0]: https://github.com/jakewarren/haveibeenpwned/compare/v1.0.0...v2.0.0
[1.0.0]: https://github.com/jakewarren/haveibeenpwned/releases/tag/v1.0.0

api/api_test.go
//nolint:funlen
package api

import (
	"testing"

	"github.com/stretchr/testify/assert"

	gock "gopkg.in/h2non/gock.v1"
)

func TestLookupEmailBreaches(t *testing.T) {
	defer gock.Off() // Flush pending mocks after test execution

	c := NewClient("")

	// test a successful call
	gock.New("https://haveibeenpwned.com").
		Reply(200).JSON(`[{
	"Title": "000webhost",
	"Name": "000webhost",
	"Domain": "000webhost.com",
	"BreachDate": "2015-03-01",
	"AddedDate": "2015-10-26T23:35:45Z",
	"ModifiedDate": "2015-10-26T23:35:45Z",
	"PwnCount": 13545468,
	"Description": "In approximately March 2015, the free web hosting provider <a href=\"http://www.troyhunt.com/2015/10/breaches-traders-plain-text-passwords.html\" target=\"_blank\" rel=\"noopener\">000webhost suffered a major data breach</a> that exposed over 13 million customer records. The data was sold and traded before 000webhost was alerted in October. The breach included names, email addresses and plain text passwords.",
	"DataClasses": ["Email addresses", "IP addresses", "Names", "Passwords"],
	"IsVerified": true,
	"IsFabricated": false,
	"IsSensitive": false,
	"IsActive": true,
	"IsRetired": false,
	"IsSpamList": false,
	"LogoType": "png"
}]`)

	breaches, err := c.LookupEmailBreaches("test@example.com")

	// err should be nil
	assert.Nil(t, err)

	// test a few values to make sure the JSON was decoded correctly
	gotBreach := breaches[0]
	assert.Equal(t, "000webhost", gotBreach.Title, "decoded breach title not equal")
	assert.Equal(t, int64(13545468), gotBreach.PwnCount, "decoded breach pwn count not equal")

	// test a 400 response code
	gock.New("https://haveibeenpwned.com").
		Reply(400)
	_, err = c.LookupEmailBreaches("test@example.com")
	if assert.NotNil(t, err, "testing 400 status code") {
		assert.Equal(t, "bad request - the provided email address is not acceptable", err.Error(), "testing 400 status code")
	}

	// test a 404 response code
	gock.New("https://haveibeenpwned.com").
		Reply(404)
	_, err = c.LookupEmailBreaches("test@example.com")
	if assert.NotNil(t, err, "testing 404 status code") {
		assert.Equal(t, "not found - the account could not be found at haveibeenpwned.com", err.Error(), "testing 404 status code")
	}

	// test a 429 response code
	gock.New("https://haveibeenpwned.com").
		Reply(429)
	_, err = c.LookupEmailBreaches("test@example.com")
	if assert.NotNil(t, err, "testing 429 status code") {
		assert.Equal(t, "rate limit exceeded", err.Error(), "testing 429 status code")
	}

	// test invalid json
	gock.New("https://haveibeenpwned.com").
		Reply(200).
		JSON("{this is broken JSON! ")

	_, jsonErr := c.LookupEmailBreaches("test@example.com")
	if assert.NotNil(t, jsonErr, "invalid json should produce error") {
		assert.Equal(t, "invalid character 't' looking for beginning of object key string", jsonErr.Error(), "did not receive expected JSON error")
	}
}

func TestLookupEmailPastes(t *testing.T) {
	defer gock.Off() // Flush pending mocks after test execution

	// test a successful call
	gock.New("https://haveibeenpwned.com").
		Reply(200).JSON(`[
{
"Source":"Pastebin",
"Id":"8Q0BvKD8",
"Title":"syslog",
"Date":"2014-03-04T19:14:54Z",
"EmailCount":139
},
{
"Source":"Pastie",
"Id":"7152479",
"Date":"2013-03-28T16:51:10Z",
"EmailCount":30
}
]`)

	c := NewClient("")

	pastes, err := c.LookupEmailPastes("test@example.com")

	// err should be nil
	assert.Nil(t, err)

	// test a few values to make sure the JSON was decoded correctly
	gotPaste := pastes[0]
	assert.Equal(t, "Pastebin", gotPaste.Source, "decoded paste source not equal")
	assert.Equal(t, "8Q0BvKD8", gotPaste.ID, "decoded paste id not equal")
	assert.Equal(t, "syslog", gotPaste.Title, "decoded paste title not equal")

	// test a 400 response code
	gock.New("https://haveibeenpwned.com").
		Reply(400)
	_, err = c.LookupEmailPastes("test@example.com")
	if assert.NotNil(t, err, "testing 400 status code") {
		assert.Equal(t, "bad request - the provided email address is not acceptable", err.Error(), "testing 400 status code")
	}

	// test a 404 response code
	gock.New("https://haveibeenpwned.com").
		Reply(404)
	_, err = c.LookupEmailPastes("test@example.com")
	if assert.NotNil(t, err, "testing 404 status code") {
		assert.Equal(t, "not found - the account could not be found at haveibeenpwned.com", err.Error(), "testing 404 status code")
	}

	// test a 429 response code
	gock.New("https://haveibeenpwned.com").
		Reply(429)
	_, err = c.LookupEmailPastes("test@example.com")
	if assert.NotNil(t, err, "testing 429 status code") {
		assert.Equal(t, "rate limit exceeded", err.Error(), "testing 429 status code")
	}

	// test invalid json
	gock.New("https://haveibeenpwned.com").
		Reply(200).
		JSON("{this is broken JSON! ")

	_, jsonErr := c.LookupEmailPastes("test@example.com")
	if assert.NotNil(t, jsonErr, "invalid json should produce error") {
		assert.Equal(t, "invalid character 't' looking for beginning of object key string", jsonErr.Error(), "did not receive expected JSON error")
	}
}


main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"

	"github.com/apex/log"
	"github.com/apex/log/handlers/cli"
	"github.com/jakewarren/haveibeenpwned/v2/api"
	"github.com/jinzhu/now"
	"gopkg.in/alecthomas/kingpin.v2"
)

var (
	app = kingpin.New("haveibeenpwned", "Un-official API client for haveibeenpwned.com.")

	debug      = app.Flag("debug", "print debug info").Short('d').Bool()
	filterDate = app.Flag("filter-date", "only print breaches released after specified date").Short('f').String()
	silent     = app.Flag("silent", "suppress response message, only display results").Short('s').Bool()
	envvar     = app.Flag("env", "environment variable to check for the HIBP API key").Default("HIBP_API_KEY").Short('e').String()
	key        = app.Flag("key", "HIBP API key").Short('k').String()

	email = app.Arg("email", "the email address to lookup.").Required().String()

	client *api.Client

	version = "(ﾉ☉ヮ⚆)ﾉ ⌒*:･ﾟ✧"
)

func main() {
	app.Version(version).VersionFlag.Short('V')
	app.HelpFlag.Short('h')
	app.UsageTemplate(kingpin.SeparateOptionalFlagsUsageTemplate)
	kingpin.MustParse(app.Parse(os.Args[1:]))

	log.SetHandler(cli.New(os.Stderr))
	log.SetLevel(log.ErrorLevel)

	if *silent {
		// turn off errors
		log.SetLevel(log.FatalLevel)
	}

	if *debug {
		log.SetLevel(log.DebugLevel)
	}

	if apikey, ok := os.LookupEnv(*envvar); ok {
		client = api.NewClient(apikey)
	} else {
		if len(*key) == 0 {
			log.Fatal("No API key provided")
		}

		client = api.NewClient(*key)
	}

	printBreachResults(*email)

	// sleep to respect the haveibeenpwned API rate limiting
	time.Sleep(2 * time.Second)

	printPasteResults(*email)
}

func printBreachResults(email string) {
	// query results for the email address
	breaches, err := client.LookupEmailBreaches(email)
	if err != nil {
		log.WithError(err).Errorf("error looking up breach data for %s", email)
		return
	}

	breachCount := 0
	var defResponse string

	for _, breach := range breaches {
		if *filterDate != "" {
			filterTime, err := now.Parse(*filterDate)
			if err != nil {
				log.WithError(err).Error("error parsing filter time")
			}

			releaseTime, err := time.Parse(time.RFC3339, breach.AddedDate)
			if err != nil {
				log.WithError(err).Error("error parsing released time")
			}

			if releaseTime.Before(filterTime) {
				log.Debugf("excluding %s (%s)", breach.Title, breach.AddedDate)
				continue
			}
		}
		defResponse += fmt.Sprintf("\n%s\n\tdomain:\t\t%s\n\tadded_date:\t%s\n\tbreach_date:\t%s\n", breach.Title, breach.Domain, breach.AddedDate, breach.BreachDate)
		defResponse += fmt.Sprintf("\temail_count:\t%s\n\tverified:\t%t\n", CommifyNumber(breach.PwnCount), breach.IsVerified)
		if *debug {
			defResponse += fmt.Sprintf("%#+v\n", breach)
		}

		defResponse += "\n"

		breachCount++
	}

	if !*silent {
		if *filterDate == "" {
			fmt.Printf("%d breaches returned for %s\n", breachCount, email)
		} else {
			fmt.Printf("%d breaches returned for %s (%d filtered out)\n", breachCount, email, (len(breaches) - breachCount))
		}
	} else if breachCount > 0 {
		fmt.Printf("%d breaches returned for %s (%d filtered out)\n", breachCount, email, (len(breaches) - breachCount))
	}

	fmt.Print(defResponse)
}

func printPasteResults(email string) {
	// query results for the email address
	pastes, err := client.LookupEmailPastes(email)
	if err != nil {
		log.WithError(err).Errorf("error looking up paste data for %s", email)
		return
	}

	pasteCount := 0
	var defResponse string

	for _, paste := range pastes {
		if *filterDate != "" {
			filterTime, err := now.Parse(*filterDate)
			if err != nil {
				log.WithError(err).Error("error parsing filter time")
			}

			releaseTime, err := time.Parse(time.RFC3339, paste.Date)
			if err != nil {
				log.WithError(err).Error("error parsing released time")
			}

			if releaseTime.Before(filterTime) {
				log.Debugf("excluding %s (%s)", paste.ID, paste.Date)
				continue
			}
		}
		defResponse += fmt.Sprintf("\n%s\n\ttitle:\t\t%s\n\tID:\t\t%s\n\tbreach_date:\t%s\n\temail_count:\t%s\n", paste.Source, paste.Title, paste.ID, paste.Date, CommifyNumber(paste.EmailCount))

		if *debug {
			defResponse += fmt.Sprintf("%#+v\n", paste)
		}

		pasteCount++
	}

	if !*silent {
		if *filterDate == "" {
			fmt.Printf("%d pastes returned for %s\n", pasteCount, email)
		} else {
			fmt.Printf("%d pastes returned for %s (%d filtered out)\n", pasteCount, email, (len(pastes) - pasteCount))
		}
	}

	fmt.Print(defResponse)
}

// CommifyNumber takes a number and returns a string with the number using comma separators
func CommifyNumber(n int64) string {
	in := strconv.FormatInt(n, 10)
	out := make([]byte, len(in)+(len(in)-2+int(in[0]/'0'))/3)
	if in[0] == '-' {
		in, out[0] = in[1:], '-'
	}

	for i, j, k := len(in)-1, len(out)-1, 0; ; i, j = i-1, j-1 {
		out[j] = in[i]
		if i == 0 {
			return string(out)
		}
		if k++; k == 3 {
			j, k = j-1, 0
			out[j] = ','
		}
	}
}


api/api.go
//nolint:funlen,dupl
package api

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

const (
	breachAPIURL = "https://haveibeenpwned.com/api/v3/breachedaccount/%s?includeUnverified=true&truncateResponse=false"
	pasteAPIURL  = "https://haveibeenpwned.com/api/v3/pasteaccount/%s"
)

// Breaches contains the details for each breach
type Breaches struct {
	AddedDate    string   `json:"AddedDate"`
	BreachDate   string   `json:"BreachDate"`
	DataClasses  []string `json:"DataClasses"`
	Description  string   `json:"Description"`
	Domain       string   `json:"Domain"`
	IsActive     bool     `json:"IsActive"`
	IsFabricated bool     `json:"IsFabricated"`
	IsRetired    bool     `json:"IsRetired"`
	IsSensitive  bool     `json:"IsSensitive"`
	IsSpamList   bool     `json:"IsSpamList"`
	IsVerified   bool     `json:"IsVerified"`
	LogoType     string   `json:"LogoType"`
	Name         string   `json:"Name"`
	PwnCount     int64    `json:"PwnCount"`
	Title        string   `json:"Title"`
}

// Pastes contains the details for each past
type Pastes struct {
	Date       string `json:"Date"`
	EmailCount int64  `json:"EmailCount"`
	ID         string `json:"Id"`
	Source     string `json:"Source"`
	Title      string `json:"Title,omitempty"`
}

// Client stores the API key
type Client struct {
	APIKey string
}

// NewClient creates a new client object configured with our API key
func NewClient(key string) *Client {
	return &Client{APIKey: key}
}

// LookupEmailBreaches returns the breach information for a given email address from HIBP.
func (c *Client) LookupEmailBreaches(email string) (breaches []Breaches, err error) {
	//lookup the breach data for the email
	endpoint := fmt.Sprintf(breachAPIURL, url.QueryEscape(email))
	client := &http.Client{
		Timeout: 15 * time.Second,
	}

	req, err := http.NewRequest("GET", endpoint, nil)
	if err != nil {
		return nil, err
	}

	//add user agent to respect API spec
	req.Header.Set("User-Agent", "haveibeenpwned")

	// add API key
	req.Header.Set("hibp-api-key", c.APIKey)

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	//check for errors from haveibeenpwned
	switch resp.StatusCode {
	case 400:
		return nil, errors.New("bad request - the provided email address is not acceptable")
	case 403:
		return nil, errors.New("forbidden — no user agent has been specified in the request")
	case 404:
		return nil, errors.New("not found - the account could not be found at haveibeenpwned.com")
	case 429:
		return nil, errors.New("rate limit exceeded")
	}

	dec := json.NewDecoder(resp.Body)
	if err = dec.Decode(&breaches); err != nil {
		return nil, err
	}

	return breaches, nil
}

// LookupEmailPastes returns the pastes from haveibeenpwned for a given
func (c *Client) LookupEmailPastes(email string) (pastes []Pastes, err error) {
	//lookup the paste data for the email
	endpoint := fmt.Sprintf(pasteAPIURL, url.QueryEscape(email))
	client := &http.Client{
		Timeout: 15 * time.Second,
	}

	req, err := http.NewRequest("GET", endpoint, nil)
	if err != nil {
		return nil, err
	}

	//add user agent to respect API spec
	req.Header.Set("User-Agent", "haveibeenpwned")

	// add API key
	req.Header.Set("hibp-api-key", c.APIKey)

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	//check for errors from haveibeenpwned
	switch resp.StatusCode {
	case 400:
		return nil, errors.New("bad request - the provided email address is not acceptable")
	case 403:
		return nil, errors.New("forbidden — no user agent has been specified in the request")
	case 404:
		return nil, errors.New("not found - the account could not be found at haveibeenpwned.com")
	case 429:
		return nil, errors.New("rate limit exceeded")
	}

	dec := json.NewDecoder(resp.Body)
	if err = dec.Decode(&pastes); err != nil {
		return nil, err
	}

	return pastes, nil
}
